### ISSUE TITLE
ts.createWatchProgram config overridden by the defaults

### ISSUE THREAD

**Issue by Matt-Esch** [typescript-eslint/typescript-eslint#4446]
**Repro**

I'm using an eslint plugin called [with-tsc-error](https://www.npmjs.com/package/eslint-plugin-with-tsc-error), which depends on the program created by typescript-estree. My project file contains the options

```
	"compilerOptions": {
		"noEmit": true,
		"incremental": true,
		"allowJs": true,
		"noUnusedLocals": true,
		"noUnusedParameters": false,
		"noImplicitAny": true
	}
```

**Expected Result**

The watch program created by the call to `ts.createWatchProgram` should have the config from my project config file.

**Actual Result**

The watch program is created with the config supplied by typescript-estree (the defaults).

**Additional Info**

It is unclear to me which module is at fault for this bug. Either typescript-estree is supplying the right thing and this is a bug in typescript, or the defaults actually mean config overrides. I opened an issue on typescript here: https://github.com/microsoft/TypeScript/issues/47476

**Versions**

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/typescript-estree` | `5.9.0` |
| `TypeScript`                           | `4.5.4` |
| `node`                                 | `12.22.6` |

**Comment by bradzacher** [comment]
Could you create a repro which shows the bug? It's a bit hard to look into this abstractly.

What is the config you are seeing? "The defaults" isn't overly descriptive.

More information would help us look into this.

**Comment by Matt-Esch** [comment]
There is a bit more info in the linked issues, however let me state it more clearly here.

I'm using a plugin called `with-tsc-error`, the purpose of which is to report on failures based on the underlying typescript config.

So I have a tscheck.json file which configures the following:

```
{
	"extends": "./tsconfig",
	"compilerOptions": {
		"noEmit": true,
		"incremental": true,
		"allowJs": true,
		"noUnusedLocals": true,
		"noUnusedParameters": false,
		"noImplicitAny": true
	}
}
```

When I run `tsc -p ./tscheck.json` it errors for noUnusedLocals and noImplicitAny and not for noUnusedParamters. The goal is to get eslint to report these errors instead of having to run the `tsc -p tscheck.json` manually. This is what `with-tsc-error` does.

So I'm setting specific compiler options for the purpose of linting. I have configured my eslint project to use the plugin and configured the parser to read my project file.

```
env:
  es6: true
  node: true
parser:
  "@typescript-eslint/parser"
parserOptions:
  sourceType: "module"
  project: "tscheck.json"
plugins:
  - "@typescript-eslint"
  - "eslint-plugin-import"
  - "unicorn"
  - "with-tsc-error"
```

The intended outcome is that eslint will report errors for unused locals and not unused parameters. The problem I'm seeing is that it is reporting errors for unused parameters. Eslint is reporting errors for what it supplies as defaults for the configuration which is 

```
optionsToExtend: {
    allowJs:true,
    allowNonTsExtensions:true,
    checkJs:true,
    noEmit:true,
    noUnusedLocals:true,
    noUnusedParameters:true
}
```

I read the source code of the module I'm using and noticed that it is fairly benign, it simply uses the provided program created by typescript-eslint to run the typescript checks. https://github.com/mkusaka/eslint-plugin-with-tsc-error/blob/main/src/rules/all.ts#L60

I then traced this through to typescript-eslint to see how this program was created.

![image](https://user-images.githubusercontent.com/1058839/149754080-37afa85c-a129-464e-a354-2a50303004f7.png)

We can see that it calls into `ts.createWatchProgram` where the `watchCompilerHost` contains these defaults I mentioned earlier. However stepping into typescript we see that typescript extends the config but gives precedence to the `watchCompilerHost.

I can't figure out who is wrong. Either this is a bug in typescript, where the supplied `host.optionsToExtend` is supposed to act as defaults rather than overrides, or this is a bug in typescript-eslint, where typescript-eslint is providing these values for `optionsToExtend` assuming they are defaults, when in fact they are actually overrides (or perhaps some other reason for forcing the config). Either way, the call to `ts.createWatchProgram` should result in a program that is configured from my configuration files and it currently ends up being configured with the defaults supplied by `typescript-eslint`.

**Comment by bradzacher** [comment]
> it errors for noUnusedLocals and noImplicitAny and not for noUnusedParamters

We set both `noUnusedLocals` and `noUnusedParamters` to on in our parser.
https://github.com/typescript-eslint/typescript-eslint/blob/f4016c24f9023e8a42def9501b68c4a908cbfede/packages/typescript-estree/src/create-program/shared.ts#L14-L32

This was done on purpose because we *used* to have a rule (`no-unused-vars-experimental`) which used TS's unused variable checks to report in ESLint. However it was VERY, VERY slow, so we removed this in the last major. Didn't bother to remove the config override because it was forgotten about and nobody has complained..

FWIW - I strongly recommend against using such a plugin. It seems like a good idea, but it's really not because it's so slow. At best it takes 50ms per file, at worst we saw triple that, or even longer. Doesn't sound like much but when you're doing that calculation on ***EVERY*** keystroke in your IDE - it burns CPU cycles like nobody's business. Additionally in the CLI it compounds across all files to destroy your lint times. 150ms * 100 files = 15 extra seconds on your lint run, and ofc scales linearly. If you use `--fix` then each file can be linted up to 11 times in a run, so worst case slowdown is `O(1.65s * files in workspace)`.

Instead of using the TS compiler options - why not just use [`no-unused-vars`](https://typescript-eslint.io/rules/no-unused-vars/), which will do the same thing without any perf problems.
